File: host/src/models/root.ts
/**
 * Where cached model files live. `GAME_BOX_MODELS_DIR` wins; otherwise the
 * platform cache directory, under `game-box/models`.
 */
import { join } from 'node:path'

export function defaultRoot(): string {
  const override = envPath('GAME_BOX_MODELS_DIR')
  if (override) return override
  const base = envPath('XDG_CACHE_HOME') ?? envPath('LOCALAPPDATA') ?? withHome() ?? '.'
  return join(base, 'game-box', 'models')
}

function withHome(): string | undefined {
  const home = envPath('HOME')
  return home === undefined ? undefined : join(home, '.cache')
}

function envPath(key: string): string | undefined {
  const value = process.env[key]
  return value ? value : undefined
}

File: models/src/root.rs
//! Where cached model files live. `GAME_BOX_MODELS_DIR` wins; otherwise the
//! platform cache directory, under `game-box/models`.

use std::path::PathBuf;

pub fn default_root() -> PathBuf {
    if let Some(dir) = env_path("GAME_BOX_MODELS_DIR") {
        return dir;
    }
    let base = env_path("XDG_CACHE_HOME")
        .or_else(|| env_path("LOCALAPPDATA"))
        .or_else(|| env_path("HOME").map(|h| h.join(".cache")))
        .unwrap_or_else(|| PathBuf::from("."));
    base.join("game-box").join("models")
}

fn env_path(key: &str) -> Option<PathBuf> {
    std::env::var_os(key).filter(|v| !v.is_empty()).map(PathBuf::from)
}
